File: chap2/sec2.3/examplePararealLorenzFig.m
sigma=10;r=28;b=8/3;                                % Lorenz rhs
f=@(t,x) [sigma*(x(2)-x(1)) r*x(1)-x(2)-x(1)*x(3) x(1)*x(2)-b*x(3)];
MF=10; MG=1;                                        % F and G time steps
F=@(t0,t1,u0) SForwardEuler(f,t0,t1,u0,MF);         % fine solver F
G=@(t0,t1,u0) SForwardEuler(f,t0,t1,u0,MG);         % coarse solver G
K=20; u0=[20;5;-5]; T=5; N=500;                          
U=Parareal(F,G,T,u0,N,K);                                
[t,u]=ForwardEuler(f,[0 T],u0,MF*N);                % fine solution
TT=0:T/N:T;
for k=1:K                                           % plot fine
  plot3(u(:,1),u(:,2),u(:,3),'-b'...                % solution and
    ,U{k}(:,1),U{k}(:,2),U{k}(:,3),'.');            % parareal iterate
  axis([-20 30 -30 40 -10 60]); view([-13,8]);
  xlabel('x'); ylabel('y'); zlabel('z');
  set(gca,'Fontsize',18)
  grid on 
  pause
end

for k=1:K+1                             
  err(k)=max(max(abs(u(1:10:end,:)-U{k})))
end      

semilogy(0:K,err,'-')
set(gca,'Fontsize',12)
xlabel('k')          
ylabel('error')
legend('parareal error')
print -depsc PararealLorenzError.eps

File: chap2/sec2.3/examplePararealLorenzFig.py
import numpy as np
import matplotlib.pyplot as plt

from ForwardEuler import forwardEuler
from Parareal import parareal

# Lorentz parameters and functions
u0 = [20., 5., -5.]
sigma, rho, beta = 10., 28., 8/3

f = lambda t, u : np.array([
    sigma*(u[1]-u[0]),
    u[0]*(rho-u[2])-u[1],
    u[0]*u[1]-beta*u[2]
    ])

MF = 10; MG = 1                                                     # F and G time steps
F = lambda t0,t1,u0: forwardEuler(f, [t0, t1], u0, MF)[-1][-1]      # fine solver F
G = lambda t0,t1,u0: forwardEuler(f, [t0, t1], u0, MG)[-1][-1]      # coarse solver F
K = 20; u0 = [20, 5, -5]; T = 5; N = 500                            # Parareal parameters
U = parareal(F, G, T, u0, N, K)                                     # solve with Parareal
t, u = forwardEuler(f,[0, T], u0, MF*N);                            # fine solution
TT = np.linspace(0, T, N+1)                                         # coarse time mesh

err = np.zeros(K+1)
for k in range(K):
    err[k] = np.max(np.abs(u[::MF] - U[k]))
plt.semilogy(err)
plt.xlabel("k")
plt.ylabel("error")
plt.tight_layout()
plt.savefig("PararealLorenzError.eps")
